xarray/backends/common.py

```diff
@@ -503,3 +503,43 @@ def assert_open(self):
         if not self._isopen:
             raise AssertionError('internal failure: file must be open '
                                  'if `autoclose=True` is used.')
+
+
+class PickleByReconstructionWrapper(object):
+
+    def __init__(self, opener, *args, **kwargs):
+
+        self.opener = opener
+        self.open_args = args
+        self.open_kwargs = kwargs
+
+        self._ds = None
+        self._isopen = False
+
+    @property
+    def value(self):
+        if self._ds is not None and self._isopen:
+            return self._ds
+        self._ds = self.opener(*self.open_args, **self.open_kwargs)
+        self._isopen = True
+        return self._ds
+
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        del state['_ds']
```

Comment:
shouldn't we close the file here? ``state['_ds'].close()``

Reply:
The basic idea is that most open file handles can't be pickled so we need to provide a mechanism to remove the existing handle and generate a new one in the dump/load steps in the pickling. In most cases, we do want to keep the original file open.